torch/distributions/constraints.py

```diff
@@ -185,21 +187,46 @@ class _LowerTriangular(Constraint):
     Constrain to lower-triangular square matrices.
     """
     def check(self, value):
-        return (torch.tril(value) == value).min(-1)[0].min(-1)[0]
+        masked_value = value * torch.tril(value.new(*value.shape[-2:]).fill_(1.0))
+        return (masked_value == value).min(-1)[0].min(-1)[0]
 
 
 class _LowerCholesky(Constraint):
     """
     Constrain to lower-triangular square matrices with positive diagonals.
     """
     def check(self, value):
+        masked_value = value * torch.tril(value.new(*value.shape[-2:]).fill_(1.0))
+        lower_triangular = (masked_value == value).min(-1)[0].min(-1)[0]
+
         n = value.size(-1)
         diag_mask = torch.eye(n, n, out=value.new(n, n))
-        lower_triangular = (torch.tril(value) == value).min(-1)[0].min(-1)[0]
         positive_diagonal = (value * diag_mask > (diag_mask - 1)).min(-1)[0].min(-1)[0]
         return lower_triangular & positive_diagonal
 
 
+class _PositiveDefinite(Constraint):
+    """
+    Constrain to positive-definite matrices.
+    """
+    def check(self, value):
+        matrix_shape = value.shape[-2:]
+        batch_shape = value.unsqueeze(0).shape[:-2]
+        # TODO: replace with batched linear algebra routine when one becomes available
+        # note that `symeig()` returns eigenvalues in ascending order
+        flattened_value = value.contiguous().view((-1,) + matrix_shape)
+        return torch.stack([v.symeig()[0][:1] > 0.0 for v in flattened_value]).view(batch_shape)
+
+
+class _RealVector(Constraint):
+    """
+    Constrain to real-valued vectors. This is the same as `constraints.real`,
+    but additionally reduces across the `event_shape` dimension.
+    """
+    def check(self, value):
+        return (value == value).min(-1)[0]
```

Comment:
nit: can you please add a comment that it fails for NaNs (it's not always obvious that `x == x` can ever be false)